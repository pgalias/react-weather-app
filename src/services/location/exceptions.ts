export enum LocalizationExceptionType {
  CANNOT_RESOLVE_LOCALIZATION = 'cannot_resolve_localization',
  OUT_TO_DATE_BROWSER = 'out_to_date_browser',
}

export class LocalizationException extends Error {
  constructor(public type: LocalizationExceptionType, message?: string) {
    super(message);
  }

  static cannotResolveLocalization(message?: string): LocalizationException {
    return new LocalizationException(
      LocalizationExceptionType.CANNOT_RESOLVE_LOCALIZATION,
      message,
    );
  }

  static outToDateBrowser(message?: string): LocalizationException {
    return new LocalizationException(LocalizationExceptionType.OUT_TO_DATE_BROWSER, message);
  }
}
