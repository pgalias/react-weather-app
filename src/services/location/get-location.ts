import { Location } from '../../models';

export const getLocation = (): Promise<Location> =>
  new Promise((resolve, reject) => {
    if (!navigator.geolocation) {
      reject();
    }

    navigator.geolocation.getCurrentPosition(
      ({ coords: { latitude, longitude } }) => resolve({ latitude, longitude }),
      () => reject(),
    );
  });
