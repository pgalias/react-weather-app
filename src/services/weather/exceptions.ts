export enum WeatherExceptionType {
  INCORRECT_DATA = 'incorrect_data',
  INTERNAL_SERVER_ERROR = 'internal_server_error',
  UNKNOWN_ERROR = 'unknown_error',
}

export class WeatherException extends Error {
  constructor(public type: WeatherExceptionType, message?: string) {
    super(message);
  }

  static incorrectData(message?: string): WeatherException {
    return new WeatherException(WeatherExceptionType.INCORRECT_DATA, message);
  }

  static internalServerError(message?: string): WeatherException {
    return new WeatherException(WeatherExceptionType.INTERNAL_SERVER_ERROR, message);
  }

  static unknownError(message?: string): WeatherException {
    return new WeatherException(WeatherExceptionType.UNKNOWN_ERROR, message);
  }
}
