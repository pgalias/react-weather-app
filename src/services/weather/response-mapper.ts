import { AxiosResponse } from 'axios';
import { Units } from '../../models';
import { Forecast } from '../../models/forecast';
import { keysToCamelCase } from '../../utils';
import { WeatherSymbols } from '../../models/weather-symbol';

type RawUnits = typeof Units[keyof typeof Units];

type Summary = {
  symbol_code: WeatherSymbols;
};

type Details = Partial<Record<RawUnits, number>>;

export type Timeserie = {
  time: string;
  data: {
    instant: {
      details: Details;
    };
    next_1_hours?: {
      summary: Summary;
    };
  };
};

export interface Response {
  properties: {
    meta: {
      units: Partial<Record<RawUnits, string>>;
    };
    timeseries: Timeserie[];
  };
}

export const responseMapper = ({ data }: AxiosResponse<Response>): Forecast => {
  const first = data.properties.timeseries[0];
  const currentWeather = {
    time: new Date(first.time),
    details: keysToCamelCase(first.data.instant.details),
    symbolCode: first.data.next_1_hours?.summary?.symbol_code ?? undefined,
  };

  return {
    units: keysToCamelCase(data.properties.meta.units),
    current: currentWeather,
  };
};
