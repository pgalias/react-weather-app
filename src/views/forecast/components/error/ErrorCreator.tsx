import React, { FunctionComponent } from 'react';
import { LocalizationExceptionType } from '../../../../services/location/exceptions';
import { WeatherExceptionType } from '../../../../services/weather/exceptions';
import { Error } from '../../../../components/error/Error';
import { errorResolver } from './error-resolver';

type Props = {
  error: LocalizationExceptionType | WeatherExceptionType;
};

export const ErrorCreator: FunctionComponent<Props> = ({ error }) => (
  <Error message={errorResolver(error)} />
);
