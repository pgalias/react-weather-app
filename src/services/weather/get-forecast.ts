import axios, { AxiosError } from 'axios';
import { Forecast, Location } from '../../models';
import { responseMapper } from './response-mapper';
import { WeatherException } from './exceptions';

export const API_URL = 'https://api.met.no/weatherapi/locationforecast/2.0/compact';

export const getForecast = (location: Location): Promise<Forecast> =>
  axios
    .get(API_URL, {
      headers: {
        Accept: 'application/json',
      },
      params: {
        lat: location.latitude,
        lon: location.longitude,
      },
    })
    .then(responseMapper)
    .catch((error: AxiosError) => {
      if (error.response?.status && error.response?.status >= 500) {
        throw WeatherException.internalServerError();
      }

      if (error.response?.status && error.response?.status >= 400) {
        throw WeatherException.incorrectData(error.response?.data?.toString());
      }

      throw WeatherException.unknownError();
    });
