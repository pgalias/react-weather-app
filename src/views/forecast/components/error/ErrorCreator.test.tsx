import React from 'react';
import { render, screen } from '@testing-library/react';
import { ErrorCreator } from './ErrorCreator';
import { LocalizationExceptionType } from '../../../../services/location/exceptions';
import { WeatherExceptionType } from '../../../../services/weather/exceptions';

describe('ErrorCreator', () => {
  test.each([
    LocalizationExceptionType.OUT_TO_DATE_BROWSER,
    LocalizationExceptionType.CANNOT_RESOLVE_LOCALIZATION,
    WeatherExceptionType.UNKNOWN_ERROR,
    WeatherExceptionType.INCORRECT_DATA,
    WeatherExceptionType.INTERNAL_SERVER_ERROR,
  ])('should render correct message for %s error', error => {
    render(<ErrorCreator error={error} />);
    expect(screen.getByRole('alert')).toMatchSnapshot();
  });
});
