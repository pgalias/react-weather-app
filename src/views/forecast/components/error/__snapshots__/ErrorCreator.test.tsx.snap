// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorCreator should render correct message for cannot_resolve_localization error 1`] = `
<div
  role="alert"
>
  Your browser or system seems to be blocking geolocation fetching. Please check the necessary permissions.
</div>
`;

exports[`ErrorCreator should render correct message for incorrect_data error 1`] = `
<div
  role="alert"
>
  Passed latitude and/or longitude may be incorrect. Please check them and try again.
</div>
`;

exports[`ErrorCreator should render correct message for internal_server_error error 1`] = `
<div
  role="alert"
>
  Cannot fetch forecast data now. Please try again in a minute.
</div>
`;

exports[`ErrorCreator should render correct message for out_to_date_browser error 1`] = `
<div
  role="alert"
>
  Your browser is stale. Please download newer one. You can find the list of compatible browsers here https://developer.mozilla.org/en-US/docs/Web/API/Navigator/geolocation#browser_compatibility
</div>
`;

exports[`ErrorCreator should render correct message for unknown_error error 1`] = `
<div
  role="alert"
>
  Something goes wrong. Please try again in a minute.
</div>
`;
